File: Assignment-1/combinations.ml
(*
	@ Author: Hao Huang
	@ Usage: In a command-line terminal, type: ocamlc -o combinations combinations.ml, and then type: ./combinations 2 5
*)

open Printf

(*
	@ Print an array to stdout
	@ Param: arr - the array to be printed
*)
let rec print_array arr = 
	let pos = ref 0 in
	while !pos < Array.length arr do
		printf "%d" arr.(!pos);
		print_string " ";
		pos := !pos + 1
	done;
	print_string "\n";;

(*
	@ Calculate C(k, n) combinations on an integer from 1 to n.
	@ Param: arr - an integer array [1, 2, ..., n]
	@ Param: res - an array to store the result and it will be re-written every time when a new result is generated.
	@ Param: n - input variable
	@ Param: k - input variable
	@ Param: i - a for loop counter to drive a loop on the arr
	@ Param: r - an index on the res to keep the position of the result
*)
let rec combination arr res n k i r = 
	if k = 0 then print_array res
	else
		let spos = i in
		let epos = (n - k) in
		for j = spos to epos do
			res.(r) <- arr.(j);
			combination arr res n (k - 1) (j + 1) (r + 1);
		done;;

let main () =
	let argc = Array.length Sys.argv in
	if argc <> 3 then print_string "Please enter k and n.\n"
	else 
		let k = int_of_string Sys.argv.(1) in
	    let n = int_of_string Sys.argv.(2) in
		(* Initialize arr and allocate space for res *)
	    let arr = Array.init n (fun i -> i + 1) in
	    let res = Array.init k (fun i -> i) in
	    combination arr res n k 0 0;	(* Start recursion from the first element in arr *)
	
	exit 0;;

main ();;
	

File: Assignment-1/combinations.cpp
/*
	@ Author: Hao Huang
	@ Usage: In a command-line terminal, type: g++ -o combinations combinations.cpp, and then type: ./combinations 2 5
*/

#include <cstdlib>
#include <vector>
#include <iostream>
using namespace std;

/*
 * @ Print a vector to stdout.
 * @ Param: res - the vector to be printed
 * @ Return: void
 */
void print(const vector<int>& res){
    int len = res.size();
    for(int i = 0; i < len; i++){
        cout << res[i] << " ";
    }
    cout << endl;
}

/*
 * @ Calculate C(k, n) combinations on an integer from 1 to n.
 * @ Param: vec - an integer vector [1, 2, ..., n]
 * @ Param: res - a vector to store the result and it will be re-written every time when a new result is generated.
 * @ Param: k - input variable
 * @ Param: n - input variable
 * @ Param: iter - an iterator object to drive a loop on the vec
 * @ Param: res_idx - an index on the res to keep the position of the result
 */
void combination(const vector<int>& vec, vector<int>& res, int k, int n, vector<int>::iterator iter, int res_idx){
    if(k == 0){
        print(res);
        return;
    }

    int spos = iter - vec.begin();
	int epos = n - k;
    for(int i = spos; i <= epos; i++){
        res[res_idx] = vec[i];
        combination(vec, res, k - 1, n, ++iter, res_idx + 1);
    }
}


int main(int argc, char* argv[]){
    int k = atoi(argv[1]);
    int n = atoi(argv[2]);
  
	// Generate an integer vector from 1 to n
    vector<int> vec;
    for(int i = 0; i < n; i++){
        vec.push_back(i + 1);
    }

    vector<int> res(k);
    vector<int>::iterator iter = vec.begin();	// An iterator to drive a loop on vec
    combination(vec, res, k, n, iter, 0);

  return 0;
}
